packages/react/react/src/hooks/useLazyImage.ts

```diff
@@ -0,0 +1,111 @@
+import { noop } from '@toss/utils';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { usePreservedCallback } from './usePreservedCallback';
+
+interface Props {
+  src: string;
+  threshold?: number | number[];
+  root?: Document | Element | null;
+  rootMargin?: string;
+  onLoadComplete?: () => void;
+  onInView?: () => void;
+}
+
+/** @tossdocs-ignore */
+export function useLazyImage({
+  src,
+  rootMargin,
+  threshold,
+  root,
+  onLoadComplete: _onLoadComplete,
+  onInView: _onInView,
+}: Props) {
+  const ref = useRef<HTMLImageElement>(null);
+
+  const onInView = usePreservedCallback(_onInView ?? noop);
+  const onLoadComplete = usePreservedCallback(_onLoadComplete ?? noop);
+
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+
+  const registerLoadEventHandler = useCallback(
+    (element: HTMLImageElement) => {
+      setIsLoading(true);
+
+      element.onload = () => {
+        setIsLoading(false);
+        onLoadComplete();
+      };
+    },
+    [onLoadComplete]
+  );
+
+  const insertImageSrc = useCallback(
+    (element: HTMLImageElement) => {
+      element.src = src;
+
+      if (element.complete) {
+        onLoadComplete();
+        return;
+      }
+      registerLoadEventHandler(element);
+    },
+    [src, onLoadComplete, registerLoadEventHandler]
+  );
+
+  const intersectionAction = useCallback(
+    ([entry]: IntersectionObserverEntry[], observer: IntersectionObserver) => {
+      if (entry) {
+        if (entry.isIntersecting) {
+          const targetImgElement = entry.target as HTMLImageElement;
+
+          insertImageSrc(targetImgElement);
+
+          /**
+           * Execute additional actions when the target element is exposed in the Viewport (or the element you specified as root)
+           * For example, logging processing
+           */
+          onInView();
+
+          /**
+           * Once the image has been loaded once, unobserve its target element to prevent repeated load.
+           */
+          observer.unobserve(targetImgElement);
+        }
+      }
+    },
+    [insertImageSrc, onInView]
+  );
+
+  useEffect(() => {
+    const imgElement = ref.current;
+
+    if (!imgElement) {
+      return;
+    }
+
+    if (typeof IntersectionObserver === 'undefined') {
+      insertImageSrc(imgElement);
+      return;
+    }
+
+    if (imgElement.getAttribute('src')) {
+      console.warn('If the "src" attribute is initially in an "img" tag, lazy load is not applied.');
+      registerLoadEventHandler(imgElement);
+      return;
+    }
```

Comment:
I think this is an unnecessary guard because it could cause confusing.

If I pass an `src` 'A' to `useLazyImage`, I think I want it to load 'A', whatever `src` an element already has is.

Since the role of `useLazyImage` is to load an given image lazily, I think it should do its own role.

```typescript
const { ref } = useLazyImage({ src: 'A' });

return <img ref={ref} src='B' />
```

Reply:
And I think this kind of confusing essentially comes from the role of `useLazyImage`, to inject a `src` attribute to the image element.

I think we could make the role lighter and the code more intuitive like:
```tsx
const { ref, inView } = useInView({ root, rootMargin, threshold });
const { isLoaded } = useImageLoad(imageURL, { disabled: !inView });

return (
  <div>
    <div ref={ref} />
    {!isLoaded ? (
      <div style={{ width: 300, height: 300 }} /> // skeleton
    ) : (
      <img src={imageURL} style={{ width: 300, height: 300 }} />
    )}
  </div>
);
```

Reply:
Slash also provides [ImpressionArea](https://github.com/toss/slash/blob/fe40c2942fd1b4ddbf1dc9b34dffdffa2b91ccae/packages/react/impression-area/src/ImpressionArea.tsx#L31) and [useImpressionArea](https://github.com/toss/slash/blob/main/packages/react/impression-area/src/useImpressionRef.tsx).

So I think the most ideal situation is that we provides smaller modules as much as possible, and developers use them depending on what they need.

```tsx
const [inView, setInView] = useState(false);
const { isLoaded } = useImageLoad(imageURL, { disabled: !inView });

return (
  <ImpressionArea onImpressionStart={() => setInView(true)} onImpressionEnd={() => setInView(false)}>
    {!isLoaded ? (
      <div style={{ width: 300, height: 300 }} /> // skeleton
    ) : (
      <img src={imageURL} style={{ width: 300, height: 300 }} />
    )}
  </ImpressionArea>
);
```

Reply:
And another reason to make `useLazyImage` lighter is that it is unintuitive to start loading the image when the element appears in viewport. Actually it is implicit in the naming.

Reply:
@HoseungJang 

>I think this is an unnecessary guard because it could cause confusing.
If I pass an src 'A' to useLazyImage, I think I want it to load 'A', whatever src an element already has is.
Since the role of useLazyImage is to load a given image lazily, I think it should do its own role.

I agree with you. This situation can happen enough. 🥲
That's why I wrote a minimalistic guard code to not put images directly into the src.
I thought that code would prevent this mess.

Reply:
@HoseungJang

I was thinking that the "useLazyLoad" hook, abstracted and designed to be specific to "image" and "lazy load", was intuitive and good for usability.

But you're also right, ideally we'd like to provide smaller, more modular hooks and let the user pick and choose which ones to use.

Additionally, "ImpressionArea" is constantly watching its element. 

From an image perspective, once an image is loaded, it doesn't need to be observed anymore. From this perspective, "ImpressionArea" is not suitable at this point in time. Additional functionality is needed to make it work for this purpose.

For example, I think we need a "disabled" option to mean that we don't want to observe it anymore.

(Since "useInView" overlaps with "ImpressionArea" in many ways, we believe it would be better to improve "ImpressionArea".)

Reply:
@HoseungJang 
I don't feel that the current issues with "useLazyImage" are critical, and I've modified the code several times based on your feedback and feel it's pretty good.  

But I also agree with you. Since you're the maintainer of that repository, I'm going to go with what you think.

Keeping that pull request open for too long is not a good direction.🙇‍♂️

Reply:
> (Since "useInView" overlaps with "ImpressionArea" in many ways, we believe it would be better to improve "ImpressionArea".)

`useInView` was just an example.

> Additionally, "ImpressionArea" is constantly watching its element.
>
> From an image perspective, once an image is loaded, it doesn't need to be observed anymore. From this perspective, "ImpressionArea" is not suitable at this point in time. Additional functionality is needed to make it work for this purpose.

I think this is related to making a feature to unobserve Intersection Observer anytime, and not related to what is the role `useLazyLoad` should have.

And although an observer handler is called every time, I think it would not have too bad impact on performance. It would be important not to re-render every time:

```tsx
const [inView, appear] = useBooleanState(false);

// No re-rendering will be triggered after `appear` is called first time. (`inView` will be always true)
return <ImpressionArea onImpressionStart={appear} />;
```

I want to finish this PR as adding just an image loader.

Reply:
@HoseungJang 
I get it!

>I want to finish this PR as adding just an image loader.

Are you referring to the "useImageLoad" example you mentioned?

Reply:
Yes, that's right.